From 66577febb6bce9b1b98c12a6c8793a11aa34289d Mon Sep 17 00:00:00 2001
From: Shengjiu Wang <shengjiu.wang@nxp.com>
Date: Thu, 12 Oct 2017 14:02:33 +0800
Subject: [PATCH 2628/5242] MLK-13946-5: ARM64: dts: enable dp audio for
 imx8qm

commit  20bab4be2f3aa133bae98bd79fa762216312430f from
https://source.codeaurora.org/external/imx/linux-imx.git

Enable dp audio for imx8qm, define sai hdmi tx and rx node.
use property "procotol" to replace the "video-mode"

Signed-off-by: Shengjiu Wang <shengjiu.wang@nxp.com>
Signed-off-by: Meng Li <Meng.Li@windriver.com>
---
 .../boot/dts/freescale/fsl-imx8qm-lpddr4-arm2.dts  |   19 +++++++
 arch/arm64/boot/dts/freescale/fsl-imx8qm.dtsi      |   54 +++++++++++++++++---
 2 files changed, 67 insertions(+), 6 deletions(-)

diff --git a/arch/arm64/boot/dts/freescale/fsl-imx8qm-lpddr4-arm2.dts b/arch/arm64/boot/dts/freescale/fsl-imx8qm-lpddr4-arm2.dts
index 1059366..394d03a 100644
--- a/arch/arm64/boot/dts/freescale/fsl-imx8qm-lpddr4-arm2.dts
+++ b/arch/arm64/boot/dts/freescale/fsl-imx8qm-lpddr4-arm2.dts
@@ -85,6 +85,13 @@
 		dais = <&sai6>, <&sai7>;
 		amix-controller = <&amix>;
 	};
+
+	sound-hdmi {
+		compatible = "fsl,imx-audio-cdnhdmi";
+		model = "imx-audio-dp";
+		audio-cpu = <&sai_hdmi_tx>;
+		protocol = <1>;
+	};
 };
 
 &acm {
@@ -126,6 +133,18 @@
 	status = "okay";
 };
 
+&sai_hdmi_tx {
+	assigned-clocks =<&clk IMX8QM_ACM_HDMI_TX_SAI0_MCLK_SEL>,
+			<&clk IMX8QM_AUD_PLL0_DIV>,
+			<&clk IMX8QM_AUD_ACM_AUD_PLL_CLK0_DIV>,
+			<&clk IMX8QM_AUD_ACM_AUD_REC_CLK0_DIV>,
+			<&clk IMX8QM_AUD_SAI_HDMITX0_MCLK>;
+	assigned-clock-parents = <&clk IMX8QM_AUD_ACM_AUD_PLL_CLK0_CLK>;
+	assigned-clock-rates = <0>, <786432000>, <49152000>, <24576000>, <49152000>;
+	fsl,sai-asynchronous;
+	status = "okay";
+};
+
 &sai6 {
 	assigned-clocks = <&clk IMX8QM_ACM_SAI6_MCLK_SEL>,
 			<&clk IMX8QM_AUD_PLL0_DIV>,
diff --git a/arch/arm64/boot/dts/freescale/fsl-imx8qm.dtsi b/arch/arm64/boot/dts/freescale/fsl-imx8qm.dtsi
index 9fa6cf6..a10d1ad 100644
--- a/arch/arm64/boot/dts/freescale/fsl-imx8qm.dtsi
+++ b/arch/arm64/boot/dts/freescale/fsl-imx8qm.dtsi
@@ -1192,13 +1192,16 @@
 				<&clk IMX8QM_HDMI_PXL_DBL_CLK>,
 				<&clk IMX8QM_HDMI_VIF_CLK>,
 				<&clk IMX8QM_HDMI_APB_MUX_CSR_CLK>,
-				<&clk IMX8QM_HDMI_APB_MUX_CTRL_CLK>;
+				<&clk IMX8QM_HDMI_APB_MUX_CTRL_CLK>,
+				<&clk IMX8QM_HDMI_I2S_CLK>,
+				<&clk IMX8QM_HDMI_I2S_BYPASS_CLK>;
 		clock-names = "dig_pll", "av_pll", "clk_ipg",
 						"clk_core", "clk_pxl", "clk_pxl_mux",
 						"clk_pxl_link", "clk_hdp", "clk_phy",
 						"clk_apb", "clk_lis","clk_msi",
 						"clk_lpcg", "clk_even","clk_dbl",
-						"clk_vif", "clk_apb_csr","clk_apb_ctrl";
+						"clk_vif", "clk_apb_csr","clk_apb_ctrl",
+						"clk_i2s", "clk_i2s_bypass";
 		power-domains = <&pd_hdmi>;
 
 		port@0 {
@@ -1999,10 +2002,12 @@
 			<0x0 0x592c0000 0x0 0x10000>, /* sai0 rx */
 			<0x0 0x592d0000 0x0 0x10000>, /* sai0 tx */
 			<0x0 0x592e0000 0x0 0x10000>, /* sai1 rx */
-			<0x0 0x592f0000 0x0 0x10000>; /* sai1 tx */
+			<0x0 0x592f0000 0x0 0x10000>, /* sai1 tx */
+			<0x0 0x59320000 0x0 0x10000>, /* sai4 rx */
+			<0x0 0x59330000 0x0 0x10000>; /* sai5 tx */
 		#dma-cells = <3>;
 		shared-interrupt;
-		dma-channels = <14>;
+		dma-channels = <16>;
 		interrupts = <GIC_SPI 374 IRQ_TYPE_LEVEL_HIGH>, /* asrc0 */
 				<GIC_SPI 375 IRQ_TYPE_LEVEL_HIGH>,
 				<GIC_SPI 376 IRQ_TYPE_LEVEL_HIGH>,
@@ -2016,14 +2021,17 @@
 				<GIC_SPI 315 IRQ_TYPE_LEVEL_HIGH>, /* sai0 */
 				<GIC_SPI 315 IRQ_TYPE_LEVEL_HIGH>,
 				<GIC_SPI 317 IRQ_TYPE_LEVEL_HIGH>, /* sai1 */
-				<GIC_SPI 317 IRQ_TYPE_LEVEL_HIGH>;
+				<GIC_SPI 317 IRQ_TYPE_LEVEL_HIGH>,
+				<GIC_SPI 326 IRQ_TYPE_LEVEL_HIGH>, /* sai4 */
+				<GIC_SPI 328 IRQ_TYPE_LEVEL_HIGH>; /* sai5 */
 		interrupt-names = "edma2-chan0-rx", "edma2-chan1-rx", /* asrc0 */
 				"edma2-chan2-rx", "edma2-chan3-tx",
 				"edma2-chan4-tx", "edma2-chan5-tx",
 				"edma2-chan6-rx", "edma2-chan7-tx", /* esai0 */
 				"edma2-chan8-rx", "edma2-chan9-tx", /* spdif0 */
 				"edma2-chan12-rx", "edma2-chan13-tx", /* sai0 */
-				"edma2-chan14-rx", "edma2-chan15-tx"; /* sai1 */
+				"edma2-chan14-rx", "edma2-chan15-tx", /* sai1 */
+				"edma2-chan18-rx", "edma2-chan19-tx"; /* sai4, sai5 */
 		status = "okay";
 	};
 
@@ -2559,6 +2567,40 @@
 		power-domains = <&pd_sai0>;
 	};
 
+	sai_hdmi_rx: sai@59080000 {
+		compatible = "fsl,imx8qm-sai";
+		reg = <0x0 0x59080000 0x0 0x10000>;
+		interrupts = <GIC_SPI 325 IRQ_TYPE_LEVEL_HIGH>;
+		clocks = <&clk IMX8QM_AUD_SAI_HDMIRX0_IPG>,
+			<&clk IMX8QM_CLK_DUMMY>,
+			<&clk IMX8QM_AUD_SAI_HDMIRX0_MCLK>,
+			<&clk IMX8QM_CLK_DUMMY>,
+			<&clk IMX8QM_CLK_DUMMY>;
+		clock-names = "bus", "mclk0", "mclk1", "mclk2", "mclk3";
+		dma-names = "rx";
+		dmas = <&edma2 18 0 1>;
+		fsl,dataline = <0xf 0x0>;
+		status = "disabled";
+		power-domains = <&pd_sai4>;
+	};
+
+	sai_hdmi_tx: sai@59090000 {
+		compatible = "fsl,imx8qm-sai";
+		reg = <0x0 0x59090000 0x0 0x10000>;
+		interrupts = <GIC_SPI 327 IRQ_TYPE_LEVEL_HIGH>;
+		clocks = <&clk IMX8QM_AUD_SAI_HDMITX0_IPG>,
+			<&clk IMX8QM_CLK_DUMMY>,
+			<&clk IMX8QM_AUD_SAI_HDMITX0_MCLK>,
+			<&clk IMX8QM_CLK_DUMMY>,
+			<&clk IMX8QM_CLK_DUMMY>;
+		clock-names = "bus", "mclk0", "mclk1", "mclk2", "mclk3";
+		dma-names = "tx";
+		dmas = <&edma2 19 0 0>;
+		fsl,dataline = <0x0 0xf>;
+		status = "disabled";
+		power-domains = <&pd_sai5>;
+	};
+
 	sai6: sai@59820000 {
 		compatible = "fsl,imx8qm-sai";
 		reg = <0x0 0x59820000 0x0 0x10000>;
-- 
1.7.9.5

